merkle/proof/proof.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package proof contains helpers for constructing log Merkle tree proofs.
+package proof
+
+import (
+	"errors"
+	"fmt"
+	"math/bits"
+
+	"github.com/google/trillian/merkle/compact"
+)
+
+// Nodes contains information on how to construct a log Merkle tree proof.
+type Nodes struct {
+	// IDs contains the IDs of non-ephemeral nodes sufficient to build the proof.
+	IDs []compact.NodeID
+	// Ephem is the ID of the ephemeal node in the proof. There is always at most
+	// one in an inclusion or a consistency proof. The ephemeral node's hash is
+	// computed from the hashes of IDs[begin:end] nodes.
+	//
+	// TODO(pavelkalinnikov): Remove this field, generate only when needed.
+	Ephem compact.NodeID
+	// begin is the beginning index (inclusive) into the IDs[begin:end] subslice
+	// of the nodes comprising the ephemeral node.
+	begin int
+	// end is the ending (exclusive) index into the IDs[begin:end] subslice of
+	// the nodes comprising the ephemeral node.
+	end int
+}
+
+// Inclusion returns the information on how to fetch and construct an inclusion
+// proof for the given leaf index in a log Merkle tree of the given size.
+func Inclusion(index, size uint64) (Nodes, error) {
+	if index >= size {
+		return Nodes{}, fmt.Errorf("index %d out of bounds for tree size %d", index, size)
+	}
+	return nodes(index, 0, size), nil
+}
+
+// Consistency returns the information on how to fetch and construct a
+// consistency proof between the two given tree sizes of a log Merkle tree.
+func Consistency(size1, size2 uint64) (Nodes, error) {
+	if size1 > size2 {
+		return Nodes{}, fmt.Errorf("tree sizes %d and %d are mis-ordered", size1, size2)
```

Comment:
maybe just say `tree size1 (%d) >  size2 (%d)` ?

Reply:
Addressed in the new `merkle` repository.